x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/index/IndexResolver.java

```diff
@@ -500,48 +355,167 @@ private GetAliasesRequest createGetAliasesRequest(FieldCapabilitiesResponse resp
+                    continue;
+                }
+                Map<String, QlFieldCapability> qlFieldTypes = new HashMap<>();
+
+                // check each type
+                for (Entry<String, FieldCapabilities> typeEntry : entry.getValue().entrySet()) {
+                    String type = typeEntry.getKey();
+                    FieldCapabilities f = typeEntry.getValue();
+                    // Skip internal fields (name starting with underscore and its type reported by field_caps starts
+                    // with underscore as well). A meta field named "_version", for example, has the type named "_version".
+                    if (type.startsWith("_") && f.getType().startsWith("_")) {
+                        continue;
+                    }
+                    
+                    QlFieldCapability qlFieldCap = new QlFieldCapability(type, f.getType(), f.isSearchable(), f.isAggregatable());
+                    boolean isValidType = false;
+
+                    // in each list of indices, keep only those belonging to the current alias
+                    isValidType = maybeSetIndices(f.indices(), indices, qlFieldCap::setIndices) || isValidType;
+                    isValidType = maybeSetIndices(f.nonAggregatableIndices(), indices, qlFieldCap::setNonAggregatableIndices)
+                        || isValidType;
+                    isValidType = maybeSetIndices(f.nonSearchableIndices(), indices, qlFieldCap::setNonSearchableIndices) || isValidType;
+
+                    // another field type is applicable to this alias, so skip the current field type
+                    if (isEmpty(qlFieldCap.getIndices())
+                        && isEmpty(qlFieldCap.getNonAggregatableIndices())
+                        && isEmpty(qlFieldCap.getNonSearchableIndices())
+                        && (isEmpty(f.indices()) == false || isEmpty(f.nonAggregatableIndices()) == false
+                            || isEmpty(f.nonSearchableIndices()) == false)) {
+                        continue;
+                    }
+                    
+                    // a type that doesn't have any indices is valid by default, for example
+                    // "id": {"keyword": {"type": "keyword","searchable": false,"aggregatable": true}}
+                    if (isValidType || (isEmpty(f.indices()) && isEmpty(f.nonAggregatableIndices()) && isEmpty(f.nonSearchableIndices()))) {
+                        qlFieldTypes.put(type, qlFieldCap);
+                    }
+                }
+                
+                if (qlFieldTypes.size() > 0) {
+                    fields.put(fieldName, qlFieldTypes);
+                }
+            }
+
+            if (fields.size() > 0) {
+                qlFieldCaps.setFields(fields);
+                map.put(alias, qlFieldCaps);
+            }
+        }
+
+        return map;
+    }
+
+    private static boolean maybeSetIndices(String[] originalList, Set<String> virtualList, Consumer<String[]> setter) {
```

Comment:
The name does not indicate what this method tries to do - maybe set indices returns true?

Reply:
`maybeSetIndices` is sometimes setting the array of indices using the `setter` on the QlFieldCapability instance. But the return value is not always `true` and, also, setting the array. Sometimes (in some edge cases) it's `true` and not calling the setter.

Reply:
how about "updateIndices" ? `maybeSet` sound non-deterministic as in, this method sometimes does X, sometimes it does not :)